src/vos/vos_obj_cache.c

```diff
@@ -314,6 +316,26 @@ vos_obj_hold(struct daos_lru_cache *occ, struct vos_container *cont,
 
 	obj->obj_incarnation = obj->obj_df->vo_incarnation;
 out:
+	if ((intent == DAOS_INTENT_PUNCH || intent == DAOS_INTENT_UPDATE) &&
+	    obj->obj_df != NULL && epoch <= obj->obj_sync_epoch &&
+	    vos_dth_get() != NULL) {
+		/* If someone has synced the object against the
+		 * obj->obj_sync_epoch, then we do not allow to modify the
+		 * object with old epoch. Let's ask the caller to retry with
+		 * newer epoch.
```

Comment:
Can we just make it a pure 'dtx sync' which just flush the cached commitable transactions? I don't think we should put this incomplete barrier functionality into this 'dtx sync'. Could you remind how it'll be used?

Reply:
One of the usages is the object based consistency verification that expects after the object sync there will be no modifications against the object with old epoch. If without such barrier functionality, it is difficult to distinguish whether the 'inconsistency' is caused by concurrent modification or not, unless we retry related list/fetch/compare, but according to our former discussion, the retry will cause the verification unclean and should be avoided.

Reply:
It looks to me that it's better to do this check in update or punch handler.